dockerfiles/entrypoint/README.md

```diff
@@ -0,0 +1,352 @@
+  -v            : show version and exit
+  -V            : show version and configure options then exit
+  -t            : test configuration and exit
+  -T            : test configuration, dump it and exit
+  -q            : suppress non-error messages during configuration testing
+  -s signal     : send signal to a master process: stop, quit, reopen, reload
+  -p prefix     : set prefix path (default: /etc/nginx/)
+  -e filename   : set error log file (default: /var/log/nginx/error.log)
+  -c filename   : set configuration file (default: /etc/nginx/nginx.conf)
+  -g directives : set global directives out of configuration file
+```
+
+### Summary
+
+So to recap an `ENTRYPOINT` allows you to configure the default executable for the container which can be extended with additional `CMD` options. Doing so allows you to create custom Dockerfiles for command-line tools and your own scripts. You'll get more practice in future lectures and learn how to modify `ENTRYPOINT` and `CMD` at runtime.
+
+Resources
+
+- <https://docs.docker.com/engine/reference/builder/#entrypoint>
+- <https://docs.docker.com/engine/reference/builder/#cmd>
+
+## Lecture 2: USING ENTRYPOINT and CMD in the CLI
+
+You might be wondering if Docker provides a way to modify the `ENTRYPOINT`  without creating a new image? Maybe you want to override the `ENTRYPOINT` at runtime the same way you can override the default `CMD` of an image.
+
+The `docker run` command has an optional `--entrpoint` flag for this. An entrypoint supplied in the command-line will overwrite any `ENTRYPOINT` defined in the Dockerfile.
+
+```shell
+docker run --entrypoint IMAGE [command] [ARG...]
+```
+
+### Exercise
+
+Let's see an example that modifies both default entrypoint and command. In the previous lecture we saw how we could print nginx help text by creating a custom image. You can do the same thing in the command-line.
+
+Run:
+
+```shell
+docker run --rm --entrypoint nginx nginx:1.21.4 -h
+```
+
+### 4 Rules
+
+There's 4 rules that describe how `CMD` and `ENTRYPOINT` interact.
+
+4 Rules For `CMD` & `ENTRYPOINT`
+
+1) Dockerfile should specify at least one of the `CMD` or `ENTRYPOINT` commands.
+2) `ENTRYPOINT` should be defined when using the container as an executable.
+3) `CMD` should be used as a way of defining default arguments for an ENTRYPOINT command or for executing an ad-hoc command in a container.
+4) `CMD` will be overridden when running the container with alternative arguments.
+
+### Gotchas
+
+When you combine the `ENTRYPOINT` and the `CMD` [the resulting command may vary](https://docs.docker.com/engine/reference/builder/#understand-how-cmd-and-entrypoint-interact) based on usage of either _shell_ or _exec_ form in the Dockerfile. If you use shell form with `ENTRYPOINT`, any `CMD` or run command-line arguments will be ignored and `ENTRYPOINT` will be started as a subcommand of `/bin/sh -c` which doesn't pass Unix signals like `SIGTERM` from `docker stop <container>`.
+
+### Exercise 2
```

Comment:
I like the rickroll, but I feel like we're going too far down the signals path, which we have yet to explain them in the course. The student won't fully understand linux signals yet. It's actually worth a lecture on them in in an advanced containers section, but not here. I'd rather the example show how a script can run via entrypoint and then pass execution to the CMD.  Let's talk about it in relation to the other examples I mentioned in a previous PR Review.